### PR TITLE
PCHR-3233: Make maximum deduction base on selected time or timeframe

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview
Currently for leave requests in hours the deduction maximum amount does not take selected time into consideration, instead, it depends on the work pattern working hours amount. In this PR a new logic is introduced:

**Multiple days**
```
Default and Maximum From Deduction = Maximum From Time - Selected From Time;
Default and Maximum To Deduction   = Selected To Time  - Minimum To Time;
```

**Single day**
```
Default and Maximum Deduction = Selected To Time - Selected From Time;
```

## Technical Details

Maximum deduction value setter function is written:

```js
// @param {String} timeType from|to
function setDeductionMaximumBoundary (timeType) {
  var uiOptions = vm.uiOptions;

  // if it is a Single day, then both "from" and "to" time inputs are related to "from" date in UI
  var dateType = uiOptions.multipleDays ? timeType : 'from';
  var timeObject = uiOptions.times[dateType];

  // only in case of multiple days mode and "to" date type the deduction calculation will consider minimum time, otherwise the selected time
  var timeFrom = uiOptions.multipleDays && dateType === 'to' ? timeObject.min : uiOptions.times.from.time;

  // only in case of multiple days mode and "from" date type the deduction calculation will consider maximum time, otherwise the selected time
  var timeTo = uiOptions.multipleDays && dateType === 'from' ? timeObject.max : uiOptions.times.to.time;

  // based on the settings above the single formula will support all 3 formulas mentioned in the PR description
  var deduction = workDays[dateType].number_of_hours ? getTimeDifferenceInHours(timeFrom, timeTo) : '0';

  // then simply set the maximum deduction value
  timeObject.maxAmount = deduction;
}
```

This function is called in `initTimeAndDateInputsWatchers()` and `loadTimeRangesFromWorkPattern()`.

## Comments

The test suite `when deduction is greater than allowed` is now obsolete because the deduction does not depend on amount of hours from work pattern.

-----

✅ Manual Tests - passed
✅ Jasmine Tests - amended and passed
✅ JS distributive files - included
⏹ PHP Unit Tests - not needed
⏹ Backstop tests - not needed
⏹ CSS distributive files - not needed

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 1]
```diff
@@ -655,27 +680,33 @@ define([
...
       }
 
       return vm.request.getWorkDayForDate(convertDateToServerFormat(date))
-        .then(function (response) {
-          timeObject.maxAmount = response.number_of_hours.toString() || '0';
+        .then(function (workDay) {
+          workDays[dateType] = workDay;
+
+          enableAndSetDataToTimeInput(dateType, workDay);
+          (isSingleDayRequest && dateType === 'from') && enableAndSetDataToTimeInput('to', workDay);
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: I know it's a bit late and this is old, but why do we need to call `enableAndSetDataToTimeInput('to', workDay);` for single day requests? Aren't they hidden anyways?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 2]
```diff
@@ -655,27 +680,33 @@ define([
...
+          setDeductionMaximumBoundary(dateType);
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Can you please confirm this: I suspect `setDeductionMaximumBoundary(dateType);` is being executed twice:

- initTimes
  - loadTimeRangesFromWorkPattern
    - setDeductionMaximumBoundary (called by loadTimeRangesFromWorkPattern)
  - setDeductionMaximumBoundary (called by initTimes)

Comment (reneolivo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: In anyway, can `setDeductionMaximumBoundary` be removed from here somehow? I feel `loadTimeRangesFromWorkPattern` is doing more and more things other than loading time ranges from work pattern (something I raised before).

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 3]
```diff
@@ -469,6 +470,19 @@ define([
...
+    /**
+     * Calculates time difference in hours
+     *
+     * @param  {String} timeFrom in HH:mm format
+     * @param  {String} timeTo in HH:mm format
+     * @return {String} amount of hours, eg. '7.5'
+     */
+    function getTimeDifferenceInHours (timeFrom, timeTo) {
+      return (getMomentDateWithGivenTime(timeTo)
+        .diff(getMomentDateWithGivenTime(timeFrom), 'minutes') / 60)
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Yes, that would round 1.5 hours to 1 hour, but we need all fractions to be preserved.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 4]
```diff
@@ -549,68 +543,99 @@ define([
...
+          absenceTypeId: vm.request.type_id,
+          value: vm.selectedAbsenceType.remainder - vm.request.balance_change
+        };
+      }
+    }
+
     /**
      * Initialises and sets the "from" and "to" times
      */
     function initTimes () {
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 5]
```diff
@@ -549,68 +543,99 @@ define([
...
+      var dateTypes = ['from'];
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 6]
```diff
@@ -549,68 +543,99 @@ define([
...
-      return $q.all(['from', 'to'].map(function (type) {
-        return loadTimeRangesFromWorkPattern(type)
-          .then(function () {
-            times[type].time = extractTimeFromServerDate(request[type + '_date']);
-            times[type].amount = Math.min(request[type + '_date_amount'], times[type].maxAmount).toString();
+      (vm.uiOptions.multipleDays) && dateTypes.push('to');
+
+      return $q.all(dateTypes.map(loadTimeRangesFromWorkPattern))
+        .then(function () {
+          ['from', 'to'].forEach(function (dateType) {
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Correct, no matter what the day mode is, we need to cycle through both.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Nope, not hidden! :) https://github.com/civicrm/civihr/pull/2443

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js [hunk 7]
```diff
@@ -549,68 +543,99 @@ define([
...
+      }
+    }
+
     /**
      * Initialises and sets the "from" and "to" times
      */
     function initTimes () {
+      var dateTypes = ['from'];
       var times = vm.uiOptions.times;
       var request = vm.request;
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js]: Done.

**Comment by igorpavlov** [comment]
The PR didn't pass the QA and the issue was introduced in https://github.com/civicrm/civihr/pull/2454/commits/42cb863c504d7e91f6a2fb11db451274f2edf4d3. It passed jasmine tests, because tests were comparing values set in UI and in request object, however, it should not be like that, since request object automatically inherited the value from UI, so the test always passed no matter what happened. This was fixed and then due to the removal of `setDeductionMaximumBoundary` from `loadTimeRangesFromWorkPattern` it did not set the maximum allowed deduction value before the current value was set, so that's why it always equaled 0.

**Pull request merged**